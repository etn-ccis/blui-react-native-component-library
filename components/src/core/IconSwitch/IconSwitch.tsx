import React, { useState } from 'react';
import { StyleSheet, TouchableOpacity, ViewProps, View, I18nManager } from 'react-native';
import Animated, {
    interpolate,
    useAnimatedStyle,
    useSharedValue,
    withTiming,
    Easing,
    Extrapolation,
} from 'react-native-reanimated';
import { Icon } from '../Icon/Icon';
import { $DeepPartial } from '@callstack/react-theme-provider';
import { ExtendedTheme, useExtendedTheme } from '@brightlayer-ui/react-native-themes';

export type IconSwitchProps = ViewProps & {
    /**
     * Flag to show the icon on the handle or not
     * @default: false
     */
    showIcon?: boolean;
    /**
     * Flag to disabled the IconSwitch
     * @default: false
     */
    disabled?: boolean;
    /**
     * Flag to pass the IconSwitch value
     * @default: false
     */
    value?: boolean;
    /**
     * Callback Event handling function to handle value change
     */
    onValueChange?: (arg: boolean) => void;
    /**
     * Theme value overrides specific to this component.
     */
    theme?: $DeepPartial<ExtendedTheme>;
};

/**
 * This is a Switch component which allow us to show the check icon on ToggleOn and Close icon on ToggleOff Switch's handle.
 */
export const IconSwitch: React.FC<IconSwitchProps> = (props) => {
    const { showIcon = false, disabled = false, value = false, onValueChange, style, ...viewProps } = props;
    const theme = useExtendedTheme(props.theme);

    const [toggled, setToggled] = useState(value);
    const shareValue = useSharedValue(value ? 1 : 0);

    const onChangeToggle = (): void => {
        setToggled(!toggled);
        onValueChange?.(!toggled);
    };

    const onPressSwitch = (): void => {
        if (shareValue.value === 0) {
            shareValue.value = withTiming(1, {
                duration: 100,
                easing: Easing.bezier(0.4, 0.0, 0.2, 1),
            });
        } else {
            shareValue.value = withTiming(0, {
                duration: 100,
                easing: Easing.bezier(0.4, 0.0, 0.2, 1),
            });
        }
        onChangeToggle();
    };

    const styles = StyleSheet.create({
        track: {
            display: 'flex',
            justifyContent: 'center',
            height: 32,
            width: 52,
            backgroundColor: disabled
                ? theme.colors.sliderTrackDisabled
                : toggled
                ? theme.colors.primary
                : theme.colors.surfaceContainerHighest,
            borderColor: toggled ? undefined : disabled ? theme.colors.disabled : theme.colors.outline,
            borderWidth: toggled ? 0 : 2,
            borderRadius: 100,
        },
        handle: {
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            width: toggled ? 24 : showIcon ? 24 : 16,
            height: toggled ? 24 : showIcon ? 24 : 16,
            borderRadius: 23,
            marginHorizontal: showIcon ? 2 : 6,
            backgroundColor: disabled
                ? toggled
                    ? theme.colors.surface
                    : theme.colors.onDisabledContainer
                : toggled
                ? theme.colors.onPrimary
                : theme.colors.onBackground,
        },
    });

    const rtl = I18nManager.isRTL;

    const toggleStyles = useAnimatedStyle(
        () => ({
            transform: [
                {
                    translateX: interpolate(
                        shareValue.value,
                        [0, 1],
                        showIcon ? (rtl ? [0, -22] : [0, 22]) : rtl ? [0, -18] : [0, 18],
                        Extrapolation.CLAMP
                    ),
                },
            ],
        }),
        []
    );

    return (
        <TouchableOpacity
            disabled={disabled}
            onPress={onPressSwitch}
            activeOpacity={1}
            style={[styles.track, style]}
            {...viewProps}
        >
            <Animated.View style={[styles.handle, toggleStyles]}>
                {showIcon && (
                    <View>
                        {toggled ? (
                            <Icon
                                source={{ family: 'material', name: 'check' }}
                                color={disabled ? theme.colors.disabled : theme.colors.onBackground}
                                size={16}
                            />
                        ) : (
                            <Icon
                                source={{ family: 'material', name: 'close' }}
                                color={disabled ? theme.colors.surfaceContainerHighest : theme.colors.onPrimary}
                                size={16}
                            />
                        )}
                    </View>
                )}
            </Animated.View>
        </TouchableOpacity>
    );
};
