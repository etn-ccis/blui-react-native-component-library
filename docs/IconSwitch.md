# Icon Switch

The `<IconSwitch />` is a switch component that can optionally renders an icon in the switch's handle.

<img width="200" alt="icon switch picture" src="./images/iconswitch.png" />

## Usage

```tsx
import { IconSwitch } from '@brightlayer-ui/react-native-components';

<IconSwitch />;

<IconSwitch showIcon />;
```

### API

<div style="overflow: auto;">

| Prop Name     | Description                      | Type                          | Required | Default |
| ------------- | -------------------------------- | ----------------------------- | -------- | ------- |
| showIcon      | To display icon or not           | `boolean`                     | no       |         |
| disabled      | Flag for render disabled switch  | `boolean`                     | no       |         |
| value         | Pass state to the component      | `boolean`                     | no       |         |
| onValueChange | Callback event handling function | `Function`                    | no       |         |
| theme         | Theme value overrides            | `$DeepPartial<ExtendedTheme>` | no       |         |

</div>

Any other props will be provided to the root element ([**View**](https://reactnative.dev/docs/view)).

### Styles

You can override the internal style used by Brightlayer UI by passing a `style` prop.
