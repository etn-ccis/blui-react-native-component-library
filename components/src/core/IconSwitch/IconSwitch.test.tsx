import React from 'react';
import { cleanup } from '@testing-library/react-native';
import TestRenderer from 'react-test-renderer';
import { IconSwitch } from './IconSwitch';

describe('IconSwitch component', () => {
    afterEach(() => {
        cleanup();
    });

    test('should render with default props', () => {
        const tree = TestRenderer.create(<IconSwitch />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render with turned on state', () => {
        const tree = TestRenderer.create(<IconSwitch value />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render with turned on state and icon', () => {
        const tree = TestRenderer.create(<IconSwitch value showIcon />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render icon switch', () => {
        const tree = TestRenderer.create(<IconSwitch showIcon />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render disabled switch with turned on state', () => {
        const tree = TestRenderer.create(<IconSwitch value disabled />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render disabled switch with turned on state and icon', () => {
        const tree = TestRenderer.create(<IconSwitch value showIcon disabled />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render disabled switch ', () => {
        const tree = TestRenderer.create(<IconSwitch disabled />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render disabled switch with icon', () => {
        const tree = TestRenderer.create(<IconSwitch disabled showIcon />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
