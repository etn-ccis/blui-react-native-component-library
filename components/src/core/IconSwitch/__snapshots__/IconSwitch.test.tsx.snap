// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconSwitch component should render disabled switch  1`] = `
<View
  accessibilityState={
    {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": undefined,
      "borderColor": undefined,
      "borderRadius": 100,
      "borderWidth": 2,
      "display": "flex",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 52,
    }
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateX": 0,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": undefined,
        "borderRadius": 23,
        "display": "flex",
        "height": 16,
        "justifyContent": "center",
        "marginHorizontal": 6,
        "transform": [
          {
            "translateX": 0,
          },
        ],
        "width": 16,
      }
    }
  />
</View>
`;

exports[`IconSwitch component should render disabled switch with icon 1`] = `
<View
  accessibilityState={
    {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": undefined,
      "borderColor": undefined,
      "borderRadius": 100,
      "borderWidth": 2,
      "display": "flex",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 52,
    }
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateX": 0,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": undefined,
        "borderRadius": 23,
        "display": "flex",
        "height": 24,
        "justifyContent": "center",
        "marginHorizontal": 2,
        "transform": [
          {
            "translateX": 0,
          },
        ],
        "width": 24,
      }
    }
  >
    <View>
      <Text
        allowFontScaling={true}
        maxFontSizeMultiplier={100}
        selectable={false}
        style={
          [
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontSize": 16,
            },
            {},
            {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;

exports[`IconSwitch component should render disabled switch with turned on state 1`] = `
<View
  accessibilityState={
    {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": undefined,
      "borderColor": undefined,
      "borderRadius": 100,
      "borderWidth": 0,
      "display": "flex",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 52,
    }
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateX": 18,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(255, 251, 254, 1)",
        "borderRadius": 23,
        "display": "flex",
        "height": 24,
        "justifyContent": "center",
        "marginHorizontal": 6,
        "transform": [
          {
            "translateX": 18,
          },
        ],
        "width": 24,
      }
    }
  />
</View>
`;

exports[`IconSwitch component should render disabled switch with turned on state and icon 1`] = `
<View
  accessibilityState={
    {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": undefined,
      "borderColor": undefined,
      "borderRadius": 100,
      "borderWidth": 0,
      "display": "flex",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 52,
    }
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateX": 22,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(255, 251, 254, 1)",
        "borderRadius": 23,
        "display": "flex",
        "height": 24,
        "justifyContent": "center",
        "marginHorizontal": 2,
        "transform": [
          {
            "translateX": 22,
          },
        ],
        "width": 24,
      }
    }
  >
    <View>
      <Text
        allowFontScaling={true}
        maxFontSizeMultiplier={100}
        selectable={false}
        style={
          [
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontSize": 16,
            },
            {},
            {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;

exports[`IconSwitch component should render icon switch 1`] = `
<View
  accessibilityState={
    {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": undefined,
      "borderColor": "rgba(121, 116, 126, 1)",
      "borderRadius": 100,
      "borderWidth": 2,
      "display": "flex",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 52,
    }
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateX": 0,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(28, 27, 31, 1)",
        "borderRadius": 23,
        "display": "flex",
        "height": 24,
        "justifyContent": "center",
        "marginHorizontal": 2,
        "transform": [
          {
            "translateX": 0,
          },
        ],
        "width": 24,
      }
    }
  >
    <View>
      <Text
        allowFontScaling={true}
        maxFontSizeMultiplier={100}
        selectable={false}
        style={
          [
            {
              "color": "rgba(255, 255, 255, 1)",
              "fontSize": 16,
            },
            {},
            {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;

exports[`IconSwitch component should render with default props 1`] = `
<View
  accessibilityState={
    {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": undefined,
      "borderColor": "rgba(121, 116, 126, 1)",
      "borderRadius": 100,
      "borderWidth": 2,
      "display": "flex",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 52,
    }
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateX": 0,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(28, 27, 31, 1)",
        "borderRadius": 23,
        "display": "flex",
        "height": 16,
        "justifyContent": "center",
        "marginHorizontal": 6,
        "transform": [
          {
            "translateX": 0,
          },
        ],
        "width": 16,
      }
    }
  />
</View>
`;

exports[`IconSwitch component should render with turned on state 1`] = `
<View
  accessibilityState={
    {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "rgba(103, 80, 164, 1)",
      "borderColor": undefined,
      "borderRadius": 100,
      "borderWidth": 0,
      "display": "flex",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 52,
    }
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateX": 18,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(255, 255, 255, 1)",
        "borderRadius": 23,
        "display": "flex",
        "height": 24,
        "justifyContent": "center",
        "marginHorizontal": 6,
        "transform": [
          {
            "translateX": 18,
          },
        ],
        "width": 24,
      }
    }
  />
</View>
`;

exports[`IconSwitch component should render with turned on state and icon 1`] = `
<View
  accessibilityState={
    {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "rgba(103, 80, 164, 1)",
      "borderColor": undefined,
      "borderRadius": 100,
      "borderWidth": 0,
      "display": "flex",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 52,
    }
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateX": 22,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(255, 255, 255, 1)",
        "borderRadius": 23,
        "display": "flex",
        "height": 24,
        "justifyContent": "center",
        "marginHorizontal": 2,
        "transform": [
          {
            "translateX": 22,
          },
        ],
        "width": 24,
      }
    }
  >
    <View>
      <Text
        allowFontScaling={true}
        maxFontSizeMultiplier={100}
        selectable={false}
        style={
          [
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontSize": 16,
            },
            {},
            {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
